_posts/2024-08-07-Introducing-a-neural-sparse-two-phase-algorithm.md

```diff
@@ -0,0 +1,122 @@
+
+## The two-phase algorithm 
+
+The two-phase search algorithm operates in two stages:
+
+1. **Initial Phase:** The algorithm uses model inference to quickly select a set of candidate documents using high-scoring tokens from the query. These high-scoring tokens, which constitute a small portion of the total number of tokens, have significant weight---or relevance---allowing for a rapid identification of potentially relevant documents. This process significantly reduces the number of documents that need to be processed, thereby lowering computational costs.
+
+2. **Recalculation Phase:** The algorithm then recalculates the scores for the candidate documents selected in the first phase, this time including both high-scoring and low-scoring tokens from the query. Although low-scoring tokens carry less weight individually, they provide valuable information in a comprehensive evaluation, particularly when long-tail terms contribute significantly to the overall score. This allows the algorithm to determine final document scores with greater accuracy.
+
+By processing documents in stages, this approach reduces computational overhead while mainitaining accuracy. The rapid selection in the first phase enhances efficiency, while the more detailed scoring in the second phase ensures accuracy. Even when handling a large number of long-tail terms, the results remain of high quality, with a notable improvement in computational efficiency.
+
+## Performance metrics
+
+We measured the speed and quality of search results using neural sparse search.
+
+### Test environment
+
+Performance was measured on OpenSearch clusters with 3 m5.4xlarge nodes using [OpenSearch Benchmark](https://opensearch.org/docs/latest/benchmark/). The tests were conducted with 20 simultaneous clients, 50 warmup iterations, and 200 test iterations.
+
+### Test dataset
+
+For search quality, we tested multiple BEIR datasets and measured the relative quality of the results. The following table presents parameter details for these datasets.
+
+| Dataset        | BEIR-Name     | Queries | Corpus | Rel D/Q |
+|----------------|---------------|---------|--------|---------|
+| NQ             | nq            | 3,452   | 2.68M  | 1.2     |
+| HotpotQA       | hotpotqa      | 7,405   | 5.23M  | 2       |
+| DBPedia        | dbpedia-entity| 400     | 4.63M  | 38.2    |
+| FEVER          | fever         | 6,666   | 5.42M  | 1.2     |
+| Climate-FEVER  | climate-fever | 1,535   | 5.42M  | 3       |
+
+### p99 latency
+
+The algorithm maintains the same inference time cost as the existing neural sparse search. To provide a clearer comparison of acceleration in the search phase, we excluded the inference step from latency calculations because inference is significantly affected by hardware type. The latency benchmark provided in this post uses raw vector search and excludes any additional impact from inference time.
```

Comment:
```suggestion
The algorithm maintains the same inference time cost as the existing neural sparse search. To provide a clearer comparison of acceleration in the search phase, we excluded the inference step from latency calculations because inference is significantly affected by hardware type. The latency benchmark provided in this post uses raw vector search and excludes any additional impact resulting from inference time.
```